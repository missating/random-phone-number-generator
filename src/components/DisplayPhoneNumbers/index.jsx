import React from 'react';
import Pagination from '../Pagination';

class DisplayPhoneNumbers extends React.Component {
  state = {
    paginatedPhoneNumbers: {},
    generatedPhoneNumbers: [],
    totalPages: 0,
    currentPage: 1,
    phoneNumbersPerPage: 10
  }

  static getPhoneNumbers = (phoneNumbers, page, phoneNumbersPerPage) => {
    const offset = (page - 1) * phoneNumbersPerPage;

    const paginatedPhoneNumbers = phoneNumbers.slice(offset).slice(0, phoneNumbersPerPage);
    const totalPages = Math.ceil(phoneNumbers.length / phoneNumbersPerPage);

    const items = {
      page,
      phoneNumbersPerPage,
      totalPages,
      nextPage: (totalPages > page) ? page + 1 : null,
      previousPage: page - 1 ? page - 1 : null,
      total: phoneNumbers.length,
      phoneNumberData: paginatedPhoneNumbers
    };

    return items;
  }

  componentDidMount() {
    const { phoneNumbers } = this.props;
    const { currentPage } = this.state;
    const paginatedPhoneNumbers = DisplayPhoneNumbers.getPhoneNumbers(phoneNumbers, currentPage, 10);
    const { totalPages } = paginatedPhoneNumbers;
    this.setState({
      paginatedPhoneNumbers,
      totalPages,
      generatedPhoneNumbers: phoneNumbers,
      currentPage
    })
  }

  static getDerivedStateFromProps(props, state) {
    const { phoneNumbers, hasGeneratedNumbers } = props;
    const { currentPage } = state;
    if (hasGeneratedNumbers) {
      const paginatedPhoneNumbers = DisplayPhoneNumbers.getPhoneNumbers(phoneNumbers, currentPage, 10);
      const { totalPages } = paginatedPhoneNumbers;
      return {
        paginatedPhoneNumbers,
        totalPages,
        phoneNumbers,
        currentPage
      }
    }

    return null;
  }

  onPageChange = (data) => {
    const { selected } = data;
    const { generatedPhoneNumbers } = this.state;
    const currentPage = Math.ceil(selected) + 1;
    const paginatedPhoneNumbers = DisplayPhoneNumbers.getPhoneNumbers(generatedPhoneNumbers, currentPage, 10);
    const { totalPages } = paginatedPhoneNumbers;
    this.setState({
      paginatedPhoneNumbers,
      totalPages,
      generatedPhoneNumbers,
      currentPage
    })
  }

  render() {
    const { paginatedPhoneNumbers, totalPages } = this.state;
    return (
      <div className="container mt-5 d-flex justify-content-center">
        <table className="table table-bordered" style={{ width: "50%" }}>
          <thead>
            {
              (paginatedPhoneNumbers.phoneNumberData && paginatedPhoneNumbers.phoneNumberData.length > 0) &&
              (
                <tr>
                  <th>#</th>
                  <th className="text-center">Generated Phone Numbers</th>
                </tr>
              )
            }
          </thead>
          <tbody className="scroll_body">
            {
              (
                paginatedPhoneNumbers.phoneNumberData && paginatedPhoneNumbers.phoneNumberData.length > 0) &&
              paginatedPhoneNumbers.phoneNumberData.map((phoneNumbers, index) => (
                <tr
                  key={`tableRow${index}`} // eslint-disable-line
                >
                  <td>{index + 1}</td>
                  <td className="text-center">{phoneNumbers.randomPhoneNumber}</td>
                </tr>
              ))
            }
          </tbody>
        </table>
        {
          (paginatedPhoneNumbers.phoneNumberData && paginatedPhoneNumbers.phoneNumberData.length < 1) &&
          (
            <div className="card" style={{ width: "18rem" }}>
              <div className="card-body">
                <h5 className="card-title">Welcome,</h5>
                <p className="card-text">You have not Generated Phone Numbers!</p>
              </div>
            </div>
          )
        }
        {
          paginatedPhoneNumbers.phoneNumberData && paginatedPhoneNumbers.phoneNumberData.length >= 1
          && (
            <Pagination
              pages={totalPages}
              onPageChange={this.onPageChange}
            />
          )
        }
      </div>
    )
  }
}

export default DisplayPhoneNumbers;
