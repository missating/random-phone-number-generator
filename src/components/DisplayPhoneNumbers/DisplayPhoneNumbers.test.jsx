// react libraries
import React from 'react';

// thired party libraries
import { shallow } from 'enzyme';

// components
import DisplayPhoneNumbers from './';

describe('DisplayPhoneNumbers Component', () => {
  it('should mount correctly', () => {
    const props = {
      phoneNumbers: [
        { randomPhoneNumber: 223344 },
        { randomPhoneNumber: 556677 },
        { randomPhoneNumber: 889900 },
      ]
    };

    const wrapper = shallow(<DisplayPhoneNumbers {...props} />);

    expect(wrapper).toMatchSnapshot()
  });

  it('should call the onPageChange methiod correctly', () => {
    const props = {
      phoneNumbers: [
        { randomPhoneNumber: 223344 },
        { randomPhoneNumber: 556677 },
        { randomPhoneNumber: 889900 },
      ]
    };

    const phoneNumbers = [
      { randomPhoneNumber: 223344 },
      { randomPhoneNumber: 556677 },
      { randomPhoneNumber: 889900 },
    ]
    const wrapper = shallow(<DisplayPhoneNumbers {...props} />);
    const instance = wrapper.instance();
    wrapper.setState({ generatedPhoneNumbers: phoneNumbers });
    const data = {
      selected: ''
    }
    instance.onPageChange(data);
    expect(wrapper.state().generatedPhoneNumbers).toEqual(phoneNumbers)
  });

  it('should get derived state from props', () => {
    const props = {
      phoneNumbers: [
        { randomPhoneNumber: 223344 },
        { randomPhoneNumber: 556677 },
        { randomPhoneNumber: 889900 },
      ],
      hasGeneratedNumbers: true
    };
    const state = {  // eslint-disable-line
      currentPage: 1
    };

    const wrapper = shallow(<DisplayPhoneNumbers {...props} />);
    expect(wrapper.state().currentPage).toEqual(1);
  });
});
