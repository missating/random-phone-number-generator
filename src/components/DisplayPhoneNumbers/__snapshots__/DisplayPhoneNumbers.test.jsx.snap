// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayPhoneNumbers Component should mount correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DisplayPhoneNumbers
    phoneNumbers={
      Array [
        Object {
          "randomPhoneNumber": 223344,
        },
        Object {
          "randomPhoneNumber": 556677,
        },
        Object {
          "randomPhoneNumber": 889900,
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <table
          className="table table-bordered"
          style={
            Object {
              "width": "50%",
            }
          }
        >
          <thead>
            <tr>
              <th>
                #
              </th>
              <th
                className="text-center"
              >
                Generated Phone Numbers
              </th>
            </tr>
          </thead>
          <tbody
            className="scroll_body"
          >
            <tr>
              <td>
                1
              </td>
              <td
                className="text-center"
              >
                223344
              </td>
            </tr>
            <tr>
              <td>
                2
              </td>
              <td
                className="text-center"
              >
                556677
              </td>
            </tr>
            <tr>
              <td>
                3
              </td>
              <td
                className="text-center"
              >
                889900
              </td>
            </tr>
          </tbody>
        </table>,
        false,
        <Pagination
          onPageChange={[Function]}
          pages={1}
        />,
      ],
      "className": "container mt-5 d-flex justify-content-center",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <thead>
              <tr>
                <th>
                  #
                </th>
                <th
                  className="text-center"
                >
                  Generated Phone Numbers
                </th>
              </tr>
            </thead>,
            <tbody
              className="scroll_body"
            >
              <tr>
                <td>
                  1
                </td>
                <td
                  className="text-center"
                >
                  223344
                </td>
              </tr>
              <tr>
                <td>
                  2
                </td>
                <td
                  className="text-center"
                >
                  556677
                </td>
              </tr>
              <tr>
                <td>
                  3
                </td>
                <td
                  className="text-center"
                >
                  889900
                </td>
              </tr>
            </tbody>,
          ],
          "className": "table table-bordered",
          "style": Object {
            "width": "50%",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <tr>
                <th>
                  #
                </th>
                <th
                  className="text-center"
                >
                  Generated Phone Numbers
                </th>
              </tr>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <th>
                    #
                  </th>,
                  <th
                    className="text-center"
                  >
                    Generated Phone Numbers
                  </th>,
                ],
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "#",
                  },
                  "ref": null,
                  "rendered": "#",
                  "type": "th",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Generated Phone Numbers",
                    "className": "text-center",
                  },
                  "ref": null,
                  "rendered": "Generated Phone Numbers",
                  "type": "th",
                },
              ],
              "type": "tr",
            },
            "type": "thead",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <tr>
                  <td>
                    1
                  </td>
                  <td
                    className="text-center"
                  >
                    223344
                  </td>
                </tr>,
                <tr>
                  <td>
                    2
                  </td>
                  <td
                    className="text-center"
                  >
                    556677
                  </td>
                </tr>,
                <tr>
                  <td>
                    3
                  </td>
                  <td
                    className="text-center"
                  >
                    889900
                  </td>
                </tr>,
              ],
              "className": "scroll_body",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": "tableRow0",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <td>
                      1
                    </td>,
                    <td
                      className="text-center"
                    >
                      223344
                    </td>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 1,
                    },
                    "ref": null,
                    "rendered": 1,
                    "type": "td",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 223344,
                      "className": "text-center",
                    },
                    "ref": null,
                    "rendered": 223344,
                    "type": "td",
                  },
                ],
                "type": "tr",
              },
              Object {
                "instance": null,
                "key": "tableRow1",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <td>
                      2
                    </td>,
                    <td
                      className="text-center"
                    >
                      556677
                    </td>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 2,
                    },
                    "ref": null,
                    "rendered": 2,
                    "type": "td",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 556677,
                      "className": "text-center",
                    },
                    "ref": null,
                    "rendered": 556677,
                    "type": "td",
                  },
                ],
                "type": "tr",
              },
              Object {
                "instance": null,
                "key": "tableRow2",
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <td>
                      3
                    </td>,
                    <td
                      className="text-center"
                    >
                      889900
                    </td>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 3,
                    },
                    "ref": null,
                    "rendered": 3,
                    "type": "td",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": 889900,
                      "className": "text-center",
                    },
                    "ref": null,
                    "rendered": 889900,
                    "type": "td",
                  },
                ],
                "type": "tr",
              },
            ],
            "type": "tbody",
          },
        ],
        "type": "table",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "onPageChange": [Function],
          "pages": 1,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <table
            className="table table-bordered"
            style={
              Object {
                "width": "50%",
              }
            }
          >
            <thead>
              <tr>
                <th>
                  #
                </th>
                <th
                  className="text-center"
                >
                  Generated Phone Numbers
                </th>
              </tr>
            </thead>
            <tbody
              className="scroll_body"
            >
              <tr>
                <td>
                  1
                </td>
                <td
                  className="text-center"
                >
                  223344
                </td>
              </tr>
              <tr>
                <td>
                  2
                </td>
                <td
                  className="text-center"
                >
                  556677
                </td>
              </tr>
              <tr>
                <td>
                  3
                </td>
                <td
                  className="text-center"
                >
                  889900
                </td>
              </tr>
            </tbody>
          </table>,
          false,
          <Pagination
            onPageChange={[Function]}
            pages={1}
          />,
        ],
        "className": "container mt-5 d-flex justify-content-center",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <thead>
                <tr>
                  <th>
                    #
                  </th>
                  <th
                    className="text-center"
                  >
                    Generated Phone Numbers
                  </th>
                </tr>
              </thead>,
              <tbody
                className="scroll_body"
              >
                <tr>
                  <td>
                    1
                  </td>
                  <td
                    className="text-center"
                  >
                    223344
                  </td>
                </tr>
                <tr>
                  <td>
                    2
                  </td>
                  <td
                    className="text-center"
                  >
                    556677
                  </td>
                </tr>
                <tr>
                  <td>
                    3
                  </td>
                  <td
                    className="text-center"
                  >
                    889900
                  </td>
                </tr>
              </tbody>,
            ],
            "className": "table table-bordered",
            "style": Object {
              "width": "50%",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <tr>
                  <th>
                    #
                  </th>
                  <th
                    className="text-center"
                  >
                    Generated Phone Numbers
                  </th>
                </tr>,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    <th>
                      #
                    </th>,
                    <th
                      className="text-center"
                    >
                      Generated Phone Numbers
                    </th>,
                  ],
                },
                "ref": null,
                "rendered": Array [
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "#",
                    },
                    "ref": null,
                    "rendered": "#",
                    "type": "th",
                  },
                  Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "host",
                    "props": Object {
                      "children": "Generated Phone Numbers",
                      "className": "text-center",
                    },
                    "ref": null,
                    "rendered": "Generated Phone Numbers",
                    "type": "th",
                  },
                ],
                "type": "tr",
              },
              "type": "thead",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <tr>
                    <td>
                      1
                    </td>
                    <td
                      className="text-center"
                    >
                      223344
                    </td>
                  </tr>,
                  <tr>
                    <td>
                      2
                    </td>
                    <td
                      className="text-center"
                    >
                      556677
                    </td>
                  </tr>,
                  <tr>
                    <td>
                      3
                    </td>
                    <td
                      className="text-center"
                    >
                      889900
                    </td>
                  </tr>,
                ],
                "className": "scroll_body",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": "tableRow0",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <td>
                        1
                      </td>,
                      <td
                        className="text-center"
                      >
                        223344
                      </td>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 1,
                      },
                      "ref": null,
                      "rendered": 1,
                      "type": "td",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 223344,
                        "className": "text-center",
                      },
                      "ref": null,
                      "rendered": 223344,
                      "type": "td",
                    },
                  ],
                  "type": "tr",
                },
                Object {
                  "instance": null,
                  "key": "tableRow1",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <td>
                        2
                      </td>,
                      <td
                        className="text-center"
                      >
                        556677
                      </td>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 2,
                      },
                      "ref": null,
                      "rendered": 2,
                      "type": "td",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 556677,
                        "className": "text-center",
                      },
                      "ref": null,
                      "rendered": 556677,
                      "type": "td",
                    },
                  ],
                  "type": "tr",
                },
                Object {
                  "instance": null,
                  "key": "tableRow2",
                  "nodeType": "host",
                  "props": Object {
                    "children": Array [
                      <td>
                        3
                      </td>,
                      <td
                        className="text-center"
                      >
                        889900
                      </td>,
                    ],
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 3,
                      },
                      "ref": null,
                      "rendered": 3,
                      "type": "td",
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": 889900,
                        "className": "text-center",
                      },
                      "ref": null,
                      "rendered": 889900,
                      "type": "td",
                    },
                  ],
                  "type": "tr",
                },
              ],
              "type": "tbody",
            },
          ],
          "type": "table",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "onPageChange": [Function],
            "pages": 1,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
